Demo/src/main/java/com/braintreepayments/demo/ShopperInsightsFragment.kt

```diff
@@ -31,142 +53,233 @@ class ShopperInsightsFragment : BaseFragment() {
-//        handleError(error)
-//    }
-//
-//    override fun onVenmoSuccess(venmoAccountNonce: VenmoAccountNonce) {
-//        super.onPaymentMethodNonceCreated(venmoAccountNonce)
-//
-//        val action = ShopperInsightsFragmentDirections.actionShopperInsightsFragmentToDisplayNonceFragment(
-//            venmoAccountNonce
-//        )
-//        NavHostFragment.findNavController(this).navigate(action)
-//    }
-//
-//    override fun onVenmoFailure(error: Exception) {
-//        handleError(error)
-//    }
+        payPalClient = PayPalClient(requireContext(), super.getAuthStringArg())
+
+        return inflater.inflate(R.layout.fragment_shopping_insights, container, false)
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        initializeViews(view)
+
+        actionButton.setOnClickListener { fetchShopperInsights() }
+        venmoButton.setOnClickListener { launchVenmo() }
+        payPalVaultButton.setOnClickListener { launchPayPalVault() }
+    }
+
+    private fun initializeViews(view: View) {
+        responseTextView = view.findViewById(R.id.responseTextView)
+        actionButton = view.findViewById(R.id.actionButton)
+        payPalVaultButton = view.findViewById(R.id.payPalVaultButton)
+        venmoButton = view.findViewById(R.id.venmoButton)
+        emailInput = view.findViewById(R.id.emailInput)
+        countryCodeInput = view.findViewById(R.id.countryCodeInput)
+        nationalNumberInput = view.findViewById(R.id.nationalNumberInput)
+        emailNullSwitch = view.findViewById(R.id.emailNullSwitch)
+        phoneNullSwitch = view.findViewById(R.id.phoneNullSwitch)
+
+        emailInput.editText?.setText("PR1_merchantname@personal.example.com")
+        nationalNumberInput.editText?.setText("4082321001")
+        countryCodeInput.editText?.setText("1")
+    }
+
+    override fun onResume() {
+        super.onResume()
+        handleVenmoReturnToApp()
+        handlePayPalReturnToApp()
+    }
+
+    private fun handlePayPalReturnToApp() {
+        if (this::paypalStartedPendingRequest.isInitialized) {
+            val paypalPaymentAuthResult =
+                paypalLauncher.handleReturnToAppFromBrowser(paypalStartedPendingRequest, Intent())
+            if (paypalPaymentAuthResult is PayPalPaymentAuthResult.Success) {
+                payPalClient.tokenize(paypalPaymentAuthResult) {
+                    when (it) {
+                        is PayPalResult.Success -> {
+                            val action =
+                                ShopperInsightsFragmentDirections
+                                    .actionShopperInsightsFragmentToDisplayNonceFragment(
+                                        it.nonce
+                                    )
+                            NavHostFragment.findNavController(this).navigate(action)
+                        }
+
+                        is PayPalResult.Failure -> {}
```

Comment:
Let's show error messages for failure or cancel message for both PayPal and Venmo in these methods